src/kubeobjects/daemonset_test.go

```diff
@@ -12,95 +12,74 @@ import (
-			Namespace: testNamespace,
-			Name:      testDaemonSetName,
-		},
-	}
-	fakeClient := fake.NewClient()
+	t.Run("create when not exists", func(t *testing.T) {
+		fakeClient := fake.NewClient()
+		annotations := map[string]string{AnnotationHash: "hash"}
+		daemonSet := createTestDaemonSetWithMatchLabels(daemonsetName, namespaceName, annotations, nil)
 
-	result, err := CreateOrUpdateDaemonSet(fakeClient, log, &dsBefore)
-	require.NoError(t, err)
-	assert.True(t, result)
+		created, err := CreateOrUpdateDaemonSet(fakeClient, log, &daemonSet)
 
-	dsAfter := &appsv1.DaemonSet{}
-	err = fakeClient.Get(context.TODO(), client.ObjectKey{
-		Namespace: testNamespace,
-		Name:      testDaemonSetName,
-	}, dsAfter)
-	require.NoError(t, err)
-	assert.Equal(t, dsBefore.Name, dsBefore.Name)
-	assert.Equal(t, dsBefore.Namespace, dsBefore.Namespace)
-}
+		require.NoError(t, err)
+		assert.True(t, created)
+	})
+	t.Run("update when exists and changed", func(t *testing.T) {
+		oldAnnotations := map[string]string{AnnotationHash: "old"}
+		oldDaemonSet := createTestDaemonSetWithMatchLabels(daemonsetName, namespaceName, oldAnnotations, nil)
+		newAnnotations := map[string]string{AnnotationHash: "new"}
+		newDaemonSet := createTestDaemonSetWithMatchLabels(daemonsetName, namespaceName, newAnnotations, nil)
+		fakeClient := fake.NewClient(&oldDaemonSet)
 
-func Test_CreateOrUpdateDaemonSet_Update(t *testing.T) {
-	dsBefore := appsv1.DaemonSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: testNamespace,
-			Name:      testDaemonSetName,
-			Annotations: map[string]string{
-				AnnotationHash: "old",
-			},
-		},
-	}
-	fakeClient := fake.NewClient(&dsBefore)
+		updated, err := CreateOrUpdateDaemonSet(fakeClient, log, &newDaemonSet)
 
-	dsUpdate := appsv1.DaemonSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: testNamespace,
-			Name:      testDaemonSetName,
-			Annotations: map[string]string{
-				AnnotationHash: "new",
-			},
-		},
-	}
-	result, err := CreateOrUpdateDaemonSet(fakeClient, log, &dsUpdate)
-	require.NoError(t, err)
-	assert.True(t, result)
+		require.NoError(t, err)
+		assert.True(t, updated)
+	})
+	t.Run("not update when exists and no changed", func(t *testing.T) {
+		oldAnnotations := map[string]string{AnnotationHash: "old"}
+		oldDaemonSet := createTestDaemonSetWithMatchLabels(daemonsetName, namespaceName, oldAnnotations, nil)
 
-	dsAfter := &appsv1.DaemonSet{}
-	err = fakeClient.Get(context.TODO(), client.ObjectKey{
-		Namespace: testNamespace,
-		Name:      testDaemonSetName,
-	}, dsAfter)
-	require.NoError(t, err)
+		fakeClient := fake.NewClient(&oldDaemonSet)
 
-	assert.NotNil(t, dsAfter.Annotations)
-	annotation, ok := dsAfter.Annotations[AnnotationHash]
-	assert.True(t, ok)
-	assert.Equal(t, "new", annotation)
-}
+		updated, err := CreateOrUpdateDaemonSet(fakeClient, log, &oldDaemonSet)
+		require.NoError(t, err)
+		assert.False(t, updated)
+	})
+	t.Run("recreate when exists and changed for immutable field", func(t *testing.T) {
```

Comment:
this test case was meant to actually test the change I introduced, 
but unfortunately the client created by `fake.NewClient` doesn't checks for this specific issue, 
so this passes with and without my changes 

should we keep it or not ?

Reply:
Does it also pass if a fault is introduced in your changes?
For example, if you introduced a trivial problem like dividing by zero in your change?

If the test breaks with an intentionally introduced fault, it validates your logic and should be kept.
Otherwise, the test is not necessary.